services/rsources/handler.go

```diff
@@ -106,24 +106,107 @@ func (*sourcesHandler) IncrementStats(ctx context.Context, tx *sql.Tx, jobRunId
 	return err
 }
 
-func (*sourcesHandler) AddFailedRecords(_ context.Context, _ *sql.Tx, _ string, _ JobTargetKey, _ []json.RawMessage) error {
-	return nil
+func (*sourcesHandler) AddFailedRecords(ctx context.Context, tx *sql.Tx, jobRunId string, key JobTargetKey, records []json.RawMessage) (err error) {
+	stmt, err := tx.Prepare(`insert into "rsources_failed_keys" (
+		job_run_id,
+		task_run_id,
+		source_id,
+		destination_id,
+		record_id
+	) values ($1, $2, $3, $4, $5)`)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = stmt.Close() }()
+
+	for i := range records {
+		_, err = stmt.ExecContext(
+			ctx,
+			jobRunId,
+			key.TaskRunID,
+			key.SourceID,
+			key.DestinationID,
+			records[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return
 }
 
-func (*sourcesHandler) GetFailedRecords(_ context.Context, _ *sql.Tx, _ string, _ JobFilter) (FailedRecords, error) {
-	return FailedRecords{}, nil
+func (sh *sourcesHandler) GetFailedRecords(ctx context.Context, jobRunId string, filter JobFilter) (FailedRecords, error) {
+	filterParams := []interface{}{}
+	filters := `WHERE job_run_id = $1`
+	filterParams = append(filterParams, jobRunId)
+
+	if len(filter.TaskRunID) > 0 {
+		filters += fmt.Sprintf(` AND task_run_id  = ANY ($%d)`, len(filterParams)+1)
+		filterParams = append(filterParams, pq.Array(filter.TaskRunID))
+	}
+
+	if len(filter.SourceID) > 0 {
```

Comment:
What is the use case where we might have to filter based on multiple sourceID's?

Reply:
It is part of the agreed contract. I guess, sourceID might change when a job is reconfigured